### PR TITLE
Remove StepListener

### PR DESCRIPTION
**Pull request by original-brownbear**
Addressing https://github.com/elastic/elasticsearch/pull/96610#pullrequestreview-1465088780 ... instead of just optimizing `onComplete` it seems smarter to just remove the thing now.

At this point all use of `StepListener` has become redundant. It was only used to provide the `whenComplete` short-hand that I replaced by wrapped listeners here.
The atomic boolean completion flag did not do anything functionally since `ListenableFuture` guards against double-invocation as well.

-> use `ListenableFuture` throughout to save on code and indirection.

This PR doesn't have any functional changes other than inlining the listenable future everywhere and replacing some `wrap` usage with the more efficient `delegateFailureAndWrap`. Unfortunately, that still came with a lot of formatting changes that I couldn't find a way to avoid but the cleanup is worth the noise IMO.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
I like it; I left one question.

Also we should get rid of this in `StatelessPersistedState` first.

**Review by DaveCTurner — COMMENTED** [review]
I like it; I left one question.

Also we should get rid of this in `StatelessPersistedState` first.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotIndexEventListener.java [hunk 1]
```diff
@@ -68,12 +68,12 @@ private static void ensureSnapshotIsLoaded(IndexShard indexShard) {
         final var store = indexShard.store();
         final SearchableSnapshotDirectory directory = unwrapDirectory(store.directory());
         assert directory != null;
-        final StepListener<Void> preWarmListener = new StepListener<>();
+        final ListenableActionFuture<Void> preWarmListener = new ListenableActionFuture<>();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotIndexEventListener.java]: wops not sure how that happened

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David! I'll deal with `StatelessPersistedState` in a sec :)

**Pull request merged**